core/src/main/kotlin/com/justai/jaicf/activator/selection/ActivationByConfidence.kt

```diff
@@ -0,0 +1,34 @@
+package com.justai.jaicf.activator.selection
+
+import com.justai.jaicf.context.BotContext
+import com.justai.jaicf.model.activation.Activation
+import com.justai.jaicf.model.activation.ActivationSelector
+import com.justai.jaicf.model.state.StatePath
+
+/**
+ * This activation selector tries to select an activation with the greatest confidence from all children
+ * of the current state, then from all siblings (including current state),
+ * then from all siblings of the parent (including the parent), and so on.
+ *
+ * @see com.justai.jaicf.BotEngine
+ * @see com.justai.jaicf.activator.Activator
+ * */
+class ActivationByConfidence : ActivationSelector {
+
+    /**
+     * @param botContext a current user's [BotContext]
+     * @param activations list of all available activations
+     * @return the most relevant [Activation]
+     *
+     * @see Activation
+     * @see ActivationSelector
+     */
+    override fun selectActivation(botContext: BotContext, activations: List<Activation>): Activation {
+        val first = StatePath.parse(activations.first().state!!)
```

Comment:
This implementation doesn't work too.
The previous implementation before changes relied on the fact, that the activations list is pre-sorted and the first element is always on the most relevant layer, that's why there was `takeWhile` that had taken only elements on the same level that the first is.
So the thing is you cannot rely on the `first` element.

Reply:
Should be fixed now.

Reply:
And this one doesn't work too, sorry :)
I guess something like that should work, but it needs testing
```
    override fun selectActivation(botContext: BotContext, activations: List<Activation>): Activation {
        val current = botContext.dialogContext.currentContext
        val (fromRoot, onPath) = activations.partition { it.state!!.commonPrefixWith(current) == "/" }
        val sorted = onPath.sortedByDescending { StatePath.parse(it.state!!).components.size } + fromRoot
        val first = StatePath.parse(sorted.first().state!!)
        return sorted.takeWhile {
            StatePath.parse(it.state!!).parent == first.parent
        }.maxBy { it.context.confidence }!!
    }
```
Briefly: we need to sort by context length first, but there can be global activators with a long context, so we split activations on ones that accessible directly from the root and ones that accessible from some state on the path from the current state to the root, and then sort adding all "from the root" states to the end.